# Aufgabe 5 - Convolutional Neural Networks - Forward Pass

Dieses Notebook thematisiert den Forward Pass durch ein Convolutional Neural Network.

Ziel ist es, den Forward Pass in PyTorch zu realisieren und die Ergebnisse aus den Übungsunterlagen zur reproduzieren.

<hr style="border-width: 5px">

### Vorbereitung
Wichtige Ergebnisse können während der Bearbeitung überprüft werden. Grundvoraussetzung hierfür ist, dass Sie das Paket `tui-dl4cv` <font color="#aa0000">installieren bzw. aktualisieren</font> und anschließend importieren.

Für die Installation stehen Ihnen zwei mögliche Wege zur Verfügung.

**(1) Installation direkt in diesem Notebook:**
Führen Sie den nachfolgenden Code-Block aus.

In [1]:
import sys

print(f"Automatically install package for '{sys.executable}'")
!{sys.executable} -m pip install tui-dl4cv \
    --extra-index-url "https://2022ws:xXCgQHZxxeNYchgryN7e@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple" \
    --no-cache --upgrade

Automatically install package for '/usr/bin/python3'
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://2022ws:****@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple
DEPRECATION: The HTML index page being used (https://nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple/tui-dl4cv/) is not a proper HTML 5 document. This is in violation of PEP 503 which requires these pages to be well-formed HTML 5 documents. Please reach out to the owners of this index page, and ask them to update this index page to a valid HTML 5 document. pip 22.2 will enforce this behaviour change. Discussion can be found at https://github.com/pypa/pip/issues/10825


ODER

**(2) Manuelle Installation über die Konsole:**
Öffnen Sie eine Konsole ("Anaconda Prompt" unter Windows) und führen Sie folgenden Befehl aus:
```text
pip install tui-dl4cv --extra-index-url "https://2022ws:xXCgQHZxxeNYchgryN7e@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple" --no-cache --upgrade
```

**Führen Sie abschließend folgenden Code-Block aus, um das Paket verwenden zu können.**
Während der Bearbeitung können Sie nun Ihre Ergebnisse mithilfe der Funktion `interactive_check` überprüfen. Die Funktionsaufrufe sind bereits an den entsprechenden Stellen im Notebook enthalten.

In [2]:
import tui_dl4cv.cnn

# noetige Erweiterung, damit Variablen aus diesem Notebook automatisch ueberprueft werden koennen
def interactive_check(name, **kwargs):
    tui_dl4cv.cnn.interactive_check(name, globals(), **kwargs)

from tui_dl4cv.cnn import print_tensors

<hr style="border-width: 5px">

### (e) Reproduzieren Sie die Ergebnisse der Forward Propagation in PyTorch.

---
Pakete importieren:

In [3]:
# PyTorch
import torch
import torch.nn.functional as F

---

*Eingabe und Gewichte als PyTorch Tensoren definieren:*

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende PyTorch-Funktion könnte für die Vervollständigung der Lücken hilfreich sein:
    <ul style="margin-bottom: 0px">
        <li><code style="background-color: #FAEAEA;">torch.tensor</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.tensor.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
    </ul>
</div>

In [4]:
# Eingabe `x` als Tensor der Groesse 1x1x8x8 definieren
x = torch.tensor([[[[0, 1, 1, 1, 0, 0, 0, 0],
                    [0, 1, 1, 1, 1, 1, 1, 0],
                    [0, 0, 0, 0, 0, 1, 0, 0],
                    [0, 0, 0, 0, 1, 1, 0, 0],
                    [0, 0, 0, 0, 1, 0, 0, 0],
                    [0, 0, 0, 1, 0, 0, 0, 0],
                    [0, 0, 1, 1, 0, 0, 0, 0],
                    [0, 0, 0, 1, 0, 0, 0, 0]]]],
                 dtype=torch.float32)

# Schicht 1
# Filtergewichte `w_1` als Tensor der Groesse 2x1x2x2 definieren
w_1 = torch.tensor([[[[9,-10],[0,10]]],
                    [[[-9,-17],[0,9]]]],
                   dtype=torch.float32)# bitte Code ergaenzen <---------------- [Luecke (1)]

# Biasgewichte `b_1` als Tensor der Groesse 2 definieren
b_1 = torch.tensor([0,0],dtype=torch.float32) # bitte Code ergaenzen <---------------- [Luecke (2)]


# Schicht 2
# Filtergewichte `w_2` als Tensor der Groesse 2x1x3x3 definieren
w_2= torch.tensor ([[[[1,0,0],[0,1,0],[0,-1,0]]],
                    [[[0,1,0],[0,0,2],[-1,0,1]]]],
                   dtype=torch.float32)# bitte Code ergaenzen <---------------- [Luecke (3)]

# Biasgewichte `b_2` als Tensor der Groesse 2 definieren
b_2 = torch.tensor([17,0],dtype=torch.float32)# bitte Code ergaenzen <---------------- [Luecke (4)]


# Schicht 3
# Filtergewichte `w_3` als Tensor der Groesse 2x2x1x1 definieren
w_3 = torch.tensor([[[[0]],[[2]]],
                   [[[1]],[[-2]]]],
                  dtype=torch.float32)# bitte Code ergaenzen <---------------- [Luecke (5)]

# Biasgewichte `b_3` als Tensor der Groesse 2 definieren
b_3 = torch.tensor([0,1],dtype=torch.float32)# bitte Code ergaenzen <---------------- [Luecke (6)]


# Schicht 5
# Gewichtsmatrix `w_5` als Tensor der Groesse 2x2 definieren
w_5 = torch.tensor([[1,2],[3,1]],dtype=torch.float32) # bitte Code ergaenzen <---------------- [Luecke (7)]

# Biasgewichte `b_5` als Tensor der Groesse 2 definieren
b_5 = torch.tensor([-34,-31],dtype=torch.float32)# bitte Code ergaenzen <---------------- [Luecke (8)]


# Implementierung ueberpruefen
interactive_check('weights')

Gewicht 'w_1':


Gewicht 'b_1':


Gewicht 'w_2':


Gewicht 'b_2':


Gewicht 'w_3':


Gewicht 'b_3':


Gewicht 'w_5':


Gewicht 'b_5':


---

*Netzwerk implementieren:*

Beachten Sie, dass im Forward Pass gleichzeitig alle zu berechnenden Werte ausgegeben werden sollen.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende PyTorch-Definitionen könnten für die Vervollständigung der Lücken hilfreich sein:
    <ul style="margin-bottom: 0px">
        <li><code style="background-color: #FAEAEA;">torch.nn.Module</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Module.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.Conv2d</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.Linear</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Linear.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.functional.silu</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.functional.silu.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.functional.adaptive_avg_pool2d</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.functional.adaptive_avg_pool2d.html#torch.nn.functional.adaptive_avg_pool2d" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.Tensor.view</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.Tensor.view.html#torch.Tensor.view" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.functional.softmax</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html#torch.nn.functional.softmax" target="_blank">PyTorch-Dokumentation</a>
        </li>
    </ul>
</div>

In [5]:
class CNN(torch.nn.Module):
    def __init__(self):
        super(CNN, self).__init__()

        # Schicht 1 `self.conv1` anlegen
        self.conv1 = torch.nn.Conv2d(in_channels=1, out_channels=2, kernel_size=2,
                                    stride=2,padding=0,dilation=1,groups=1,
                                    bias=True) # bitte Code ergaenzen <---------------- [Luecke (9)]

        # Schicht 2 `self.conv2` anlegen
        self.conv2 = torch.nn.Conv2d(in_channels=2, out_channels=2, kernel_size=3,
                                    stride=1,padding=1,dilation=2,groups=2,
                                    bias=True)# bitte Code ergaenzen <---------------- [Luecke (10)]

        # Schicht 3 `self.conv3` anlegen
        self.conv3 = torch.nn.Conv2d(in_channels=2, out_channels=2, kernel_size=1,
                                    stride=1,padding=0,dilation=1,groups=1,
                                    bias=True)# bitte Code ergaenzen <---------------- [Luecke (11)]

        # Schicht 5 `self.fc` anlegen
        self.fc = torch.nn.Linear(in_features=2,out_features=2,bias=True) # bitte Code ergaenzen <---------------- [Luecke (12)]

        # Gewichte mit bereits definierten Variablen initialisieren
        self.conv1.weight.data = w_1
        self.conv1.bias.data = b_1
        self.conv2.weight.data = w_2
        self.conv2.bias.data = b_2
        self.conv3.weight.data = w_3
        self.conv3.bias.data = b_3
        self.fc.weight.data = w_5
        self.fc.bias.data = b_5

    def forward(self, x):
        # Schicht 1: Aktivierung `z_1` und Ausgabe `o_1` berechnen
        z_1 = self.conv1(x) # bitte Code ergaenzen <---------------- [Luecke (13)]
        o_1 = F.silu(z_1) # bitte Code ergaenzen <---------------- [Luecke (14)]

        print_tensors(tensors=(z_1, o_1),
                      labels=('Aktivierung z_1', 'Ausgabe o_1'))

        # Schicht 2: Aktivierung `z_2` und Ausgabe `o_2` berechnen
        z_2 = self.conv2(o_1) # bitte Code ergaenzen <---------------- [Luecke (15)]
        o_2 = F.silu(z_2) # bitte Code ergaenzen <---------------- [Luecke (16)]

        print_tensors(tensors=(z_2, o_2),
                      labels=('Aktivierung z_2', 'Ausgabe o_2'))

        # Schicht 3: Aktivierung `z_3` und Ausgabe `o_3` berechnen
        z_3 = self.conv3(o_2) # bitte Code ergaenzen <---------------- [Luecke (17)]
        o_3 = F.silu(z_3) # bitte Code ergaenzen <---------------- [Luecke (18)]

        print_tensors(tensors=(z_3, o_3),
                      labels=('Aktivierung z_3', 'Ausgabe o_3'))

        # Schicht 4: Global Average Pooling
        o_4 = F.adaptive_avg_pool2d(o_3,(1,1))# bitte Code ergaenzen <---------------- [Luecke (19)]

        print_tensors(tensors=o_4,
                      labels='Ausgabe o_4')

        # Uebergang zu vollverschalteten Schichten
        o_4 = o_4.view((-1,2))# bitte Code ergaenzen <---------------- [Luecke (20)]

        print_tensors(tensors=o_4,
                      labels='Eingabe o_4 fuer vollverschaltete Schicht')

        # Schicht 5: Aktivierung `z_5` und Ausgabe `o_5` berechnen
        z_5 = self.fc(o_4) # bitte Code ergaenzen <---------------- [Luecke (21)]
        o_5 = F.softmax(z_5, dim=1) # bitte Code ergaenzen <---------------- [Luecke (22)]

        print_tensors(tensors=(z_5, o_5),
                      labels=('Aktivierung z_5', 'Ausgabe o_5'),
                      precision=3)

        return z_5

---

*Netzwerkobjekt anlegen, auf Eingabe anwenden und Kreuzentropie berechnen:*

In [6]:
# Netzwerkobjekt anlegen
network = CNN()

# Netzwerkobjekt auf Eingabe anwenden
y = network(x) # bitte Code ergaenzen <---------------- [Luecke (23)]


# Kreuzentropie berechnen
t = torch.tensor([1], dtype=torch.long)
e_ce = F.cross_entropy(y, t)

print_tensors(tensors=e_ce,
              labels="Fehlermaß Kreuzentropie",
              precision=4)

interactive_check('loss')

Aktivierung z_1:
[[[[  0.   9.  10.   0.]
   [  0.   0.   0.   0.]
   [  0.  10.   9.   0.]
   [  0.   9.   0.   0.]]

  [[ -8. -17.   9.   0.]
   [  0.   0.  -8.   0.]
   [  0.   9.  -9.   0.]
   [  0. -17.   0.   0.]]]]
Ausgabe o_1:
[[[[ 0.  9. 10.  0.]
   [ 0.  0.  0.  0.]
   [ 0. 10.  9.  0.]
   [ 0.  9.  0.  0.]]

  [[-0. -0.  9.  0.]
   [ 0.  0. -0.  0.]
   [ 0.  9. -0.  0.]
   [ 0. -0.  0.  0.]]]]
Aktivierung z_2:
[[[[ 8. 17.]
   [27. 26.]]

  [[ 0.  0.]
   [-0.  9.]]]]
Ausgabe o_2:
[[[[ 8. 17.]
   [27. 26.]]

  [[ 0.  0.]
   [-0.  9.]]]]
Aktivierung z_3:
[[[[ 0.  0.]
   [-0. 18.]]

  [[ 9. 18.]
   [28.  9.]]]]
Ausgabe o_3:
[[[[ 0.  0.]
   [-0. 18.]]

  [[ 9. 18.]
   [28.  9.]]]]
Ausgabe o_4:
[[[[ 4.5]]

  [[16. ]]]]
Eingabe o_4 fuer vollverschaltete Schicht:
[[ 4.5 16. ]]
Aktivierung z_5:
[[ 2.498 -1.504]]
Ausgabe o_5:
[[0.982 0.018]]
Fehlermaß Kreuzentropie:
4.0201


Fehlermass Kreuzentropie 'e_ce':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
    <br>
    <i>Augaben bei korrekter Implementierung:</i>
    <br>
    <code style="padding: 0">
Aktivierung z_1:
[[[[  0.   9.  10.   0.]
   [  0.   0.   0.   0.]
   [  0.  10.   9.   0.]
   [  0.   9.   0.   0.]]
  [[ -8. -17.   9.   0.]
   [  0.   0.  -8.   0.]
   [  0.   9.  -9.   0.]
   [  0. -17.   0.   0.]]]]
Ausgabe o_1:
[[[[ 0.  9. 10.  0.]
   [ 0.  0.  0.  0.]
   [ 0. 10.  9.  0.]
   [ 0.  9.  0.  0.]]
  [[-0. -0.  9.  0.]
   [ 0.  0. -0.  0.]
   [ 0.  9. -0.  0.]
   [ 0. -0.  0.  0.]]]]
Aktivierung z_2:
[[[[ 8. 17.]
   [27. 26.]]
  [[ 0.  0.]
   [-0.  9.]]]]
Ausgabe o_2:
[[[[ 8. 17.]
   [27. 26.]]
  [[ 0.  0.]
   [-0.  9.]]]]
Aktivierung z_3:
[[[[ 0.  0.]
   [-0. 18.]]
  [[ 9. 18.]
   [28.  9.]]]]
Ausgabe o_3:
[[[[ 0.  0.]
   [-0. 18.]]
  [[ 9. 18.]
   [28.  9.]]]]
Ausgabe o_4:
[[[[ 4.5]]
  [[16. ]]]]
Eingabe o_4 fuer vollverschaltete Schicht:
[[ 4.5 16. ]]
Aktivierung z_5:
[[ 2.498 -1.504]]
Ausgabe o_5:
[[0.982 0.018]]
Fehlermaß Kreuzentropie:
4.0201
</code>
</details>

$_{_\text{Created for Deep Learning for Computer Vision (DL4CV)}}$